meterfs/crypt.c

```diff
@@ -0,0 +1,79 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Meterfs cryptography
+ *
+ * Copyright 2024 Phoenix Systems
+ * Author: Aleksander Kaminski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <stdlib.h>
+#include <errno.h>
+#include <string.h>
+
+#include <board_config.h>
+
+#if METERFS_AES_ENABLE
+#include <tinyaes/aes.h>
+#endif
+
+#include "crypt.h"
+
+void meterfs_constructIV(uint8_t *iv, const file_t *f, const entry_t *e)
+{
+	static const uint8_t filler[16] = {
+		0x55, 0xaa, 0x55, 0xaa, 0x55, 0xaa, 0x55, 0xaa,
+		0x55, 0xaa, 0x55, 0xaa, 0x55, 0xaa, 0x55, 0xaa
+	};
+
+	iv[0] = (e->id.no >> 0) & 0xff;
+	iv[1] = (e->id.no >> 8) & 0xff;
+	iv[2] = (e->id.no >> 16) & 0xff;
+	iv[3] = (e->id.no >> 24) & 0xff;
+
+	iv[4] = (f->header.sector >> 0) & 0xff;
+	iv[5] = (f->header.sector >> 8) & 0xff;
+	iv[6] = (f->header.sector >> 16) & 0xff;
+	iv[7] = (f->header.sector >> 24) & 0xff;
+
+	iv[8] = (f->header.sectorcnt >> 0) & 0xff;
+	iv[9] = (f->header.sectorcnt >> 8) & 0xff;
+	iv[10] = (f->header.sectorcnt >> 16) & 0xff;
+	iv[11] = (f->header.sectorcnt >> 24) & 0xff;
+
+	iv[12] = (f->header.recordsz >> 0) & 0xff;
+	iv[13] = (f->header.recordsz >> 8) & 0xff;
+	iv[14] = (f->header.recordsz >> 16) & 0xff;
+	iv[15] = (f->header.recordsz >> 24) & 0xff;
```

Comment:
The `meterfs_constructIV()` uses some `f->header.sector`, `f->header.sectorcnt`, `f->header.recordsz` and entry `e->id.no` information as part of the initialization vector. If these values are predictable (eg. may be found in partition table or fs internal - not encrypted - structures), it can lead to weak encryption, especially when used in conjunction with the same encryption keys.
The initialization vectors should be unpredictable for each encryption, which provides an additional layer of security. The `55aa55aa..` filler pattern is easily seekable in memory so potentially if one has access to RAM can find IV.

Reply:
This is intended to secure against spi flash being read externally, that's all. I don't have anything else to use as replicable IV. I could like hash this, but it's rather security by obscurity.

IV being unique for each record should be enough to not compromise security for the intended use case

Reply:
Thanks for clarification regarding the ensured uniqueness of IV for each record. Combined with the currently used ctr mode, this should suffice to maintain security for the intended use case, despite concerns about predictability, from a security risk assessment perspective in final product.